middleware/compress_brotli.go

```diff
@@ -0,0 +1,121 @@
+package middleware
+
+import (
+	"bufio"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/andybalholm/brotli"
+
+	"github.com/labstack/echo/v4"
+)
+
+type (
+	// BrotliConfig defines the config for Brotli middleware.
+	BrotliConfig struct {
+		// Skipper defines a function to skip middleware.
+		Skipper Skipper
+
+		// Brotli compression level.
+		// Optional. Default value -1.
+		Level int `yaml:"level"`
+	}
+
+	brotliResponseWriter struct {
+		io.Writer
+		http.ResponseWriter
+	}
+)
+
+const (
+	brotliScheme = "br"
+)
+
+var (
+	// DefaultBrotliConfig is the default Brotli middleware config.
+	DefaultBrotliConfig = BrotliConfig{
+		Skipper: DefaultSkipper,
+		Level:   brotli.DefaultCompression,
+	}
+)
+
+// Brotli returns a middleware which compresses HTTP response using brotli compression
+// scheme.
+func Brotli() echo.MiddlewareFunc {
+	return BrotliWithConfig(DefaultBrotliConfig)
+}
+
+// BrotliWithConfig return Brotli middleware with config.
+// See: `Brotli()`.
+func BrotliWithConfig(config BrotliConfig) echo.MiddlewareFunc {
+	// Defaults
+	if config.Skipper == nil {
+		config.Skipper = DefaultBrotliConfig.Skipper
+	}
+	if config.Level == 0 {
+		config.Level = DefaultBrotliConfig.Level
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+
+			res := c.Response()
+			res.Header().Add(echo.HeaderVary, echo.HeaderAcceptEncoding)
+			if strings.Contains(c.Request().Header.Get(echo.HeaderAcceptEncoding), brotliScheme) {
+				res.Header().Set(echo.HeaderContentEncoding, brotliScheme) // Issue #806
+				rw := res.Writer
+
+				w := brotli.NewWriterOptions(rw, brotli.WriterOptions{Quality: config.Level})
```

Comment:
Please use a sync.Pool here to reduce the amount of allocations performed by this middleware. Yo can check the current implementation of compress.go or use #1672 as a reference